Dataflow combinator library.

horned-sphere opened the pull request:
A library of combinators to transform streams of data from Swim lanes. There are two main components:

1. A fluent API for constructing combinator graphs, [found here](https://github.com/swimos/swim/tree/feature/combinators/swim-system-java/swim-core-java/swim.dataflow/src/main/java/swim/dataflow/graph). The primary interfaces are SwimStream and MapSwimStream upon which transformations can be stacked.
2. Semantic interpretation of the combinator graphs, [found here](https://github.com/swimos/swim/tree/feature/combinators/swim-system-java/swim-core-java/swim.dataflow/src/main/java/swim/dataflow/connector). Based around the Junction/MapJunction Receptacle/MapReceptacle interfaces.

See [DataFlowAgent](https://github.com/swimos/swim/blob/feature/combinators/swim-system-java/swim-mesh-java/swim.api/src/main/java/swim/api/agent/DataflowAgent.java) for an example of how combinator graphs can be attached to agent lanes and downlinks.

c9r commented:
Excellent!

I'd like to weave a couple of the fundamental utilities into the core libraries. I think it makes sense to add `Deferred` into `swim-util` so that other orthogonal libraries can provide compatible deferred computations. `Recurring` may be a useful addition to `swim-concurrent`. I'll add some comments on the impl inline in the PR.

It might make sense to move the graph impl into a separate library to codify the decoupling from the connector interfaces. One possibility is to replace the existing `swim-streamlet` library with the dataflow connector interfaces, and elevate the graph impl to the top level of the `swim-dataflow` library. As an aside, what do others think of the streamlet name? Is it worth keeping as the library name? I wouldn't rename any of the dataflow connector interfaces to streamlet, but we could keep the library name. If anything, it seems like `SwimStream` could be a candidate to rename to `Streamlet`, which would actually be an argument for up leveling the dataflow connector interfaces to the top of the `swim-dataflow` library, and move the impl into the `swim-streamlet` library and reverse the dependency order.

I'd like to make the `DataflowAgent` functionality in a little deeper. I think I prefer your approach of explicitly wrapping lanes and links in `SwimStream`s, rather than having the lane/link interface implement `Receptacle` and `Junction` directly.

I'd like to put the actual driver for materializing and executing dataflow graphs into a kernel module, to facilitate pluggable materializers, such as one that exposes the materialized graph via introspection.

horned-sphere commented:
> I'd like to weave a couple of the fundamental utilities into the core libraries. I think it makes sense to add `Deferred` into `swim-util` so that other orthogonal libraries can provide compatible deferred computations. `Recurring` may be a useful addition to `swim-concurrent`. 

That sounds very sensible. I'll move them over.

> 
> It might make sense to move the graph impl into a separate library to codify the decoupling from the connector interfaces. One possibility is to replace the existing `swim-streamlet` library with the dataflow connector interfaces, and elevate the graph impl to the top level of the `swim-dataflow` library. 

Splitting the interface and implementation definitely makes sense. I'm not sure the way I've written it for now is very amedable to there being multiple implementations but that could be an improvement for the future.

> As an aside, what do others think of the streamlet name? Is it worth keeping as the library name? I wouldn't rename any of the dataflow connector interfaces to streamlet, but we could keep the library name. If anything, it seems like `SwimStream` could be a candidate to rename to `Streamlet`, which would actually be an argument for up leveling the dataflow connector interfaces to the top of the `swim-dataflow` library, and move the impl into the `swim-streamlet` library and reverse the dependency order.

I actually think that the opposite of what you suggest would be better. Rather than renaming the `Stream` interfaces to `Streamlet`, I think streamlet would be a better name for the runtime side (what I have called junctions and receptacles). The names I chose were to avoid collisions with things that were already there so if we get rid of the old streamlets we could rename those.

> 
> I'd like to make the `DataflowAgent` functionality in a little deeper. I think I prefer your approach of explicitly wrapping lanes and links in `SwimStream`s, rather than having the lane/link interface implement `Receptacle` and `Junction` directly.
> 
> I'd like to put the actual driver for materializing and executing dataflow graphs into a kernel module, to facilitate pluggable materializers, such as one that exposes the materialized graph via introspection.

The `DataflowAgent` was just meant to be a motivating example to show how it could be used. I wasn't very happy with it. Perhaps we could discuss how you think streams should be instantiated when you are here? I agree on not implementing the interfaces for the lanes and downlinks; there's not really any advantage over just registering them as observers.

horned-sphere commented:
I've made an attempt and doing some of the improvements Chris suggested. For now it's in another [branch](https://github.com/swimos/swim/tree/feature/newstreamlets).

1. Deleted the old streamlet interfaces.
2. Moved generally useful things into swim-util.
3. Split the dataflow project across three different modules.
4. Renamed the `Conduit` and `MapConduit` interfaces to be `Streamlet` and `MapStreamlet`.

It's done as a sequence of commits so as much or as little as desired could be merged into this.
